main.go

```diff
@@ -2,1980 +2,502 @@ package main
-			relations: relations,
+	// Process input derivations recursively
+	for inputDrvPath := range drv.InputDerivations {
+		if _, loaded := ctx.visited.LoadOrStore(inputDrvPath, true); !loaded {
+			ctx.wg.Add(1)
+			go processDerivation(inputDrvPath, "", ctx)
 		}
 	}
 }
 
-// AddRebuildInfo adds rebuild information for a FOD
-// For DBBatcher, this is a no-op as the data is already in the SQLite database
-func (b *DBBatcher) AddRebuildInfo(drvPath string, status, actualHash, errorMessage string) {
-	// No-op for DBBatcher since the rebuild data is already in the database
-	// The data will be retrieved directly from the rebuild_queue table as needed
-}
-
-func (b *DBBatcher) IncrementDrvCount() {
-	b.mu.Lock()
-	b.stats.drvs++
-	shouldLog := time.Since(b.lastStatsTime) >= 3*time.Second
-	b.mu.Unlock()
-
-	if shouldLog {
-		b.mu.Lock()
-		b.logStats()
-		b.mu.Unlock()
-	}
-}
-
-// Modify Flush to use the writer goroutine
-func (b *DBBatcher) Flush() {
-	b.mu.Lock()
-
-	if len(b.fodBatch) > 0 {
-		// Copy batches to local variables
-		fods := make([]FOD, len(b.fodBatch))
-		copy(fods, b.fodBatch)
-		relations := make([]DrvRevision, len(b.relationBatch))
-		copy(relations, b.relationBatch)
-
-		// Clear batches
-		b.fodBatch = b.fodBatch[:0]
-		b.relationBatch = b.relationBatch[:0]
-
-		b.mu.Unlock()
-
-		// Send to writer goroutine and wait for result
-		resultChan := make(chan error, 1)
-		b.writeChan <- writeOperation{
-			fods:       fods,
-			relations:  relations,
-			resultChan: resultChan,
-		}
-		<-resultChan
-	} else {
-		b.mu.Unlock()
-	}
-}
-
-// Modify Close to shut down the writer goroutine
-func (b *DBBatcher) Close() error {
-	b.Flush()
-	close(b.writeChan)
-	<-b.done // Wait for writer to finish
-
-	if err := b.fodStmt.Close(); err != nil {
-		return err
-	}
-	if err := b.relStmt.Close(); err != nil {
-		return err
-	}
-	return nil
-}
-
-// ProcessingContext holds the context for processing derivations
-type ProcessingContext struct {
-	batcher        Writer // Interface for adding FODs
-	visited        *sync.Map
-	processedPaths *sync.Map // Cache for already processed derivation paths
-	wg             *sync.WaitGroup
-	semaphore      chan struct{} // Limit concurrency
-}
-
-func processDerivation(inputFile string, ctx *ProcessingContext) {
-	ctx.batcher.IncrementDrvCount()
+// rebuildFOD attempts to rebuild a FOD to verify it works
+// rebuildFOD attempts to rebuild a FOD from source to verify it and its dependencies.
+// It explicitly disables binary caches to ensure a local, from-source build.
+func rebuildFOD(fod *FOD, ctx *ProcessingContext) {
+	log.Printf("Verifying FOD from source (no cache): %s", fod.DrvPath)
 
-	if config.Debug {
-		debugLog("PROCESS: Starting derivation %s", inputFile)
-	}
+	// Combine into a single, definitive check.
+	// We want to force a from-source build and check the result.
+	buildCmd := exec.CommandContext(ctx.ctx, "nix-build", fod.DrvPath, "--no-out-link", "--no-substitute", "--check")
```

Comment:
[`--check` requires the derivation to be already built](https://nix.dev/manual/nix/2.30/command-ref/nix-store/realise.html#options), so this will fail most of the time. When would we be doing the first run? We could of course also have a conditional that checks for existence of the outpath first.